src/torchcodec/_core/BetaCudaDeviceInterface.cpp

```diff
@@ -0,0 +1,563 @@
+  if (isFlushing_) {
+    return 0;
+  }
+
+  TORCH_CHECK(pPicParams != nullptr, "Invalid picture parameters");
+  TORCH_CHECK(decoder_, "Decoder not initialized before picture decode");
+
+  // Send frame to be decoded by NVDEC - non-blocking call.
+  CUresult result = cuvidDecodePicture(decoder_.get(), pPicParams);
+  if (result != CUDA_SUCCESS) {
+    return 0; // Yes, you're reading that right, 0 mean error.
+  }
+
+  // The frame was sent to be decoded on the NVDEC hardware. Now we store some
+  // relevant info into our frame buffer so that we can retrieve the decoded
+  // frame later when receiveFrame() is called.
+  // Importantly we need to 'guess' the PTS of that frame. The heuristic we use
+  // (like in DALI) is that the frames are ready to be decoded in the same order
+  // as the packets were sent to the parser. So we assign the PTS of the frame
+  // by popping the PTS of the oldest packet in our packetsPtsQueue (note:
+  // oldest doesn't necessarily mean lowest PTS!).
+
+  TORCH_CHECK(
+      // TODONVDEC P0 the queue may be empty, handle that.
+      !packetsPtsQueue.empty(),
+      "PTS queue is empty when decoding a frame");
+  int64_t guessedPts = packetsPtsQueue.front();
+  packetsPtsQueue.pop();
+
+  // Field values taken from DALI
+  CUVIDPARSERDISPINFO dispInfo = {};
+  dispInfo.picture_index = pPicParams->CurrPicIdx;
+  dispInfo.progressive_frame = !pPicParams->field_pic_flag;
+  dispInfo.top_field_first = pPicParams->bottom_field_flag ^ 1;
+  dispInfo.repeat_first_field = 0;
+  dispInfo.timestamp = guessedPts;
+
+  FrameBufferSlot* slot = findEmptySlot();
+  slot->dispInfo = dispInfo;
+  slot->guessedPts = guessedPts;
+  slot->occupied = true;
+
+  return 1;
+}
+
+// Moral equivalent of avcodec_receive_frame(). Here, we look for a decoded
+// frame with the exact desired PTS in our frame buffer. This logic is only
+// valid in exact seek_mode, for now.
+int BetaCudaDeviceInterface::receiveFrame(
+    UniqueAVFrame& avFrame,
+    int64_t desiredPts) {
+  FrameBufferSlot* slot = findFrameWithExactPts(desiredPts);
+  if (slot == nullptr) {
+    // No frame found, instruct caller to try again later after sending more
+    // packets.
+    return AVERROR(EAGAIN);
+  }
+
+  slot->occupied = false;
+  slot->guessedPts = -1;
+
+  CUVIDPROCPARAMS procParams = {};
+  CUVIDPARSERDISPINFO dispInfo = slot->dispInfo;
+  procParams.progressive_frame = dispInfo.progressive_frame;
+  procParams.top_field_first = dispInfo.top_field_first;
+  procParams.unpaired_field = dispInfo.repeat_first_field < 0;
+  CUdeviceptr framePtr = 0;
+  unsigned int pitch = 0;
+
+  // We know the frame we want was sent to the hardware decoder, but now we need
+  // to "map" it to an "output surface" before we can use its data. This is a
+  // blocking calls that waits until the frame is fully decoded and ready to be
+  // used.
+  CUresult result = cuvidMapVideoFrame(
+      static_cast<CUvideodecoder>(decoder_.get()),
```

Comment:
This line is what made me realize we're declaring `UniqueCUvideodecoder` with `void` instead of `CUvideodecoder`.